packages/synthetics-sdk-broken-links/src/broken_links.ts

```diff
@@ -12,24 +12,187 @@
+
+  // scrape origin_url for all links
+  // (shuffle links if necessary and) truncate at link_limit
+
+  // create new page to be used for all scraped links
+  // navigate to each link - LOOP:
+  //          each call to `checkLinks(...)` will return a `SyntheticLinkResult`
+  //          Object added to an array of `followed_links`
+
+  // returned a SyntheticResult with `options`, `followed_links` &
+  // runtimeMetadata
   return {} as SyntheticResult;
 }
+
+async function checkLink(
+  page: Page,
+  link: LinkIntermediate,
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions
+): Promise<BrokenLinksResultV1_SyntheticLinkResult> {
+  // START - to resolve warnings while under development
+  page;
+  link;
+  options;
+  // END - to resolve warnings  while under development
+
+  // PSEUDOCODE
+
+  // determine expected_status_code—determined here rather than `navigate(...)`
+  // since it will be used later when creating
+  // `BrokenLinksResultV1_SyntheticLinkResult` Object
+
+  // call `navigate(...)`
+  /**
+   * const {
+   *    response: responseOrError,
+   *    passed,
+   *    testing_only,
+   *    link_start_time,
+   *    link_end_time,
+   * } = await navigate(page, link, expected_status_code, options);
+   */
+
+  // Error handling based on `responseOrError` & passed:
+  //    if      `responseOrError` is type `Error` set `error_type` and
+  //            `error_message` in `SyntheticLinkResult` accordingly
+  //    else if `passed === false`. Set error `SYNTHETIC_INCORRECT_STATUS_CODE`
+
+  // return `SynheticLinkResult` with all calculated information
+  return {} as BrokenLinksResultV1_SyntheticLinkResult;
+}
+
+/**
+ * Navigates to a target URL with retries and timeout handling.
+ *
+ * @param page - The Puppeteer Page instance.
+ * @param link - The LinkIntermediate containing the target URL.
+ * @param expected_status_code - The expected HTTP status code.
+ * @param options - The options for navigation and retries.
+ * @returns Information about navigation attempt:
+ *   - `response`: HTTP response or error if navigation fails, or null.
+ *   - `passed`: Boolean indicating if navigation passed per status code.
+ *   - `retriesRemaining`: Remaining retries after attempt. (for testing)
+ *   - `link_start_time`: Start time of navigation attempt.
+ *   - `link_end_time`: End time of navigation attempt.
+ */
+export async function navigate(
+  page: Page,
+  link: LinkIntermediate,
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions,
+  expected_status_code: ResponseStatusCode = {
+    status_class: ResponseStatusCode_StatusClass.STATUS_CLASS_2XX,
+  }
+): Promise<{
+  response: HTTPResponse | Error | null;
+  passed: boolean;
+  retriesRemaining: number;
+  link_start_time: string;
+  link_end_time: string;
+}> {
+  let link_start_time = '';
+  let link_end_time = '';
+  let retriesRemaining = options.max_retries!;
+  const per_link_timeout_millis =
```

Comment:
All timeout things:
- User set synthetic/wtrace timeout (max 300 secs)
- GCF response timeout (hopefully <= wtrace timeout)
- global set timeout for each individual visited links
- optional setting to specify a new timeout for an individual link

`const per_link_timeout_millis =
    options.per_link_options[link.target_url]?.link_timeout_millis ||
    options.link_timeout_millis!;`
    
Calculates what timeout we should use for "this" link. `options.per_link_options[link.target_url]?.link_timeout_millis` sees if the customer has set a new timeout specific to this link.target_url. If not use the global one set, options.link_timeout_millis

Reply:
I added a comment to help explain to next reader

Reply:
communicate to customer that we retry in every error scenario (wrong status code or timeout error). Is this the conventional?

Do we even want to support max_retries? (for uptime checks we don't retry in case of failure). Ask product.

Also should max_retries be set to 0 or 1 to mean we only try once.

Reply:
default max_retries is 0